ProcessLib/ThermoHydroMechanics/ThermoHydroMechanicsFEM-impl.h

```diff
@@ -0,0 +1,522 @@
+
+namespace ProcessLib
+{
+namespace ThermoHydroMechanics
+{
+template <typename ShapeFunctionDisplacement, typename ShapeFunctionPressure,
+          typename IntegrationMethod, int DisplacementDim>
+ThermoHydroMechanicsLocalAssembler<ShapeFunctionDisplacement,
+                                   ShapeFunctionPressure, IntegrationMethod,
+                                   DisplacementDim>::
+    ThermoHydroMechanicsLocalAssembler(
+        MeshLib::Element const& e,
+        std::size_t const /*local_matrix_size*/,
+        bool const is_axially_symmetric,
+        unsigned const integration_order,
+        ThermoHydroMechanicsProcessData<DisplacementDim>& process_data)
+    : _process_data(process_data),
+      _integration_method(integration_order),
+      _element(e),
+      _is_axially_symmetric(is_axially_symmetric)
+{
+    unsigned const n_integration_points =
+        _integration_method.getNumberOfPoints();
+
+    _ip_data.reserve(n_integration_points);
+    _secondary_data.N_u.resize(n_integration_points);
+
+    auto const shape_matrices_u =
+        initShapeMatrices<ShapeFunctionDisplacement,
+                          ShapeMatricesTypeDisplacement, IntegrationMethod,
+                          DisplacementDim>(e, is_axially_symmetric,
+                                           _integration_method);
+
+    auto const shape_matrices_p =
+        initShapeMatrices<ShapeFunctionPressure, ShapeMatricesTypePressure,
+                          IntegrationMethod, DisplacementDim>(
+            e, is_axially_symmetric, _integration_method);
+
+    auto const& solid_material =
+        MaterialLib::Solids::selectSolidConstitutiveRelation(
+            _process_data.solid_materials, _process_data.material_ids,
+            e.getID());
+
+    for (unsigned ip = 0; ip < n_integration_points; ip++)
+    {
+        _ip_data.emplace_back(solid_material);
+        auto& ip_data = _ip_data[ip];
+        auto const& sm_u = shape_matrices_u[ip];
+        _ip_data[ip].integration_weight =
+            _integration_method.getWeightedPoint(ip).getWeight() *
+            sm_u.integralMeasure * sm_u.detJ;
+
+        ip_data.N_u_op = ShapeMatricesTypeDisplacement::template MatrixType<
+            DisplacementDim, displacement_size>::Zero(DisplacementDim,
+                                                      displacement_size);
+        for (int i = 0; i < DisplacementDim; ++i)
+            ip_data.N_u_op
+                .template block<1, displacement_size / DisplacementDim>(
+                    i, i * displacement_size / DisplacementDim)
+                .noalias() = sm_u.N;
+
+        ip_data.N_u = sm_u.N;
+        ip_data.dNdx_u = sm_u.dNdx;
+
+        ip_data.N_p = shape_matrices_p[ip].N;
+        ip_data.dNdx_p = shape_matrices_p[ip].dNdx;
+
+        _secondary_data.N_u[ip] = shape_matrices_u[ip].N;
+    }
+}
+
+template <typename ShapeFunctionDisplacement, typename ShapeFunctionPressure,
```

Comment:
Please add a comment here like:
Assembles the local Jacobian matrix. So far, the linearisation of HT part is not considered as that in HT process.

Reply:
The error of the ip_data comes from the PR of the update of the parameters. I am updating the parameters now

Reply:
The comment is added.

Reply:
The ip_data part is updated.